scipy/sparse/csgraph/_tools.pyx

```diff
@@ -334,16 +334,25 @@ def reconstruct_path(csgraph, predecessors, directed=True):
     pind = predecessors[indices]
     indptr = pind.searchsorted(np.arange(N + 1)).astype(ITYPE)
 
-    if directed == True:
-        data = csgraph[pind, indices]
-    else:
-        data1 = csgraph[pind, indices]
+    data = csgraph[pind, indices]
+
+    # Fix issue #4018:
+    # If `pind` and `indices` are empty arrays, `data` is a sparse matrix
+    # (it is a numpy.matrix otherwise); handle this case separately.
+    try:
+        data = data.getA1()
+    except:
```

Comment:
Please don't use a bare `except`. Any exception other than `AttributeError` could signal a bug.

In fact, I would use an explicit type check here:

```py
if isspmatrix(data):
    data = data.todense()
```